pkg/helm/frrk8s.go

```diff
@@ -0,0 +1,238 @@
+		isOpenShift:        isOCP,
+		namespace:          namespace,
+		kubeRbacProxyImage: &imageInfo{},
+	}
+	var err error
+	frrk8sImage := os.Getenv("FRRK8S_IMAGE")
+	if frrk8sImage != "" {
+		controllerRepo, controllerTag := getImageNameTag(frrk8sImage)
+		config.frrk8sImage = &imageInfo{controllerRepo, controllerTag}
+	}
+
+	frrImage := os.Getenv("FRR_IMAGE")
+	if frrImage == "" {
+		return nil, errors.Errorf("FRR_IMAGE env variable must be set for frr-k8s")
+	}
+	frrRepo, frrTag := getImageNameTag(frrImage)
+	config.frrImage = &imageInfo{frrRepo, frrTag}
+
+	config.metricsPort, err = valueWithDefault("FRRK8S_METRICS_PORT", 7572)
+	if err != nil {
+		return nil, err
+	}
+	config.secureMetricsPort, err = valueWithDefault("FRRK8S_HTTPS_METRICS_PORT", 9140)
+	if err != nil {
+		return nil, err
+	}
+
+	config.frrMetricsPort, err = valueWithDefault("FRRK8S_FRR_METRICS_PORT", 7573)
+	if err != nil {
+		return nil, err
+	}
+	config.secureFRRMetricsPort, err = valueWithDefault("FRRK8S_FRR_HTTPS_METRICS_PORT", 9141)
+	if err != nil {
+		return nil, err
+	}
+
+	// We shouldn't spam the api server trying to apply ServiceMonitors if the resource isn't installed.
+	if os.Getenv("DEPLOY_SERVICEMONITORS") == "true" {
+		config.enableServiceMonitor = true
+	}
+
+	kubeRbacProxyImage := os.Getenv("KUBE_RBAC_PROXY_IMAGE")
+	if kubeRbacProxyImage == "" {
+		return nil, errors.Errorf("KUBE_RBAC_PROXY_IMAGE env variable must be set")
+	}
+	config.kubeRbacProxyImage.repo, config.kubeRbacProxyImage.tag = getImageNameTag(kubeRbacProxyImage)
+	err = validateFRRK8SConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return config, nil
+}
+
+func validateFRRK8SConfig(c *frrK8SChartConfig) error {
+	if c.isOpenShift && !c.enableServiceMonitor {
+		return fmt.Errorf("service monitors are required on OpenShift")
+	}
+
+	return nil
+}
+
+// GetObjects retrieve manifests from chart after patching custom values passed in crdConfig
+// and environment variables.
+func (h *FRRK8SChart) GetObjects(crdConfig *metallbv1beta1.MetalLB, withPrometheus bool) ([]*unstructured.Unstructured, error) {
+	chartValueOpts := &values.Options{}
+	chartValues, err := chartValueOpts.MergeValues(getter.All(h.envSettings))
+	if err != nil {
+		return nil, err
+	}
+
+	h.config.patchToChartValues(crdConfig, withPrometheus, chartValues)
+	release, err := h.client.Run(h.chart, chartValues)
+	if err != nil {
+		return nil, err
+	}
+	objs, err := parseManifest(release.Manifest)
+	if err != nil {
+		return nil, err
+	}
+	for _, obj := range objs {
+		// Set namespace explicitly into non cluster-scoped resource because helm doesn't
+		// patch namespace into manifests at client.Run.
+		objKind := obj.GetKind()
+		if objKind != "PodSecurityPolicy" {
+			obj.SetNamespace(h.namespace)
+		}
+		// we need to override the security context as helm values are added on top
+		// of hardcoded ones in values.yaml, so it's not possible to reset runAsUser
+		if isFRRK8SWebhookDeployment(obj) && h.config.isOpenShift {
```

Comment:
I'd argue that the frrk8s webhook must be deployed upfront (similarly to what we do with metallb, so the validation of crds is available before we deploy. Need to think about this.

Reply:
well a given user may still need to have his configuration ready upfront, nad that configuration might include a frrconfiguration to accept routes (this is what we go advertising). And as we saw with metallb, they might be willing to deploy them before activating metallb.
This PR is already quite big, let's defer it to a different pr (but imo it must be done).

Reply:
fair enough, I'll work on it in a separate pr